packages/create-waku/src/helpers/example-option.ts

```diff
@@ -116,15 +118,130 @@ export async function downloadAndExtractRepo(
       cwd: root,
       strip: 2 + name.split('/').length,
-      filter: (p) => p.includes(`waku-main/examples/${name}/`),
+      filter: (p) =>
+        p.replace(/waku-[^/]+\//, '').includes(`examples/${name}/`),
     }),
   );
 }
+
+function getValidURL(url: string): URL | undefined {
+  try {
+    return new URL(url);
+  } catch (error) {
+    const err = error as Error & { code: string | undefined };
+    if (err.code !== 'ERR_INVALID_URL') {
+      console.error(error);
+      process.exit(1);
+    }
+  }
+}
+
+type ParsedExampleOption = {
+  example: string;
+  defaultRef: string;
+  repoInfo?: RepoInfo;
+};
+
+export async function parseExampleOption(
+  example: string | undefined,
+  defaultRef: string,
+): Promise<ParsedExampleOption | undefined> {
+  if (!example) {
+    return undefined;
+  }
+
+  const repoUrl = getValidURL(example);
+  if (repoUrl) {
+    // NOTE check github origin
+    if (repoUrl.origin !== 'https://github.com') {
+      console.error(
+        `Invalid URL: ${red(
+          `"${example}"`,
+        )}. Only GitHub repositories are supported. Please use a GitHub URL and try again.`,
+      );
+      process.exit(1);
+    }
+
+    const repoInfo = await getRepoInfo(repoUrl);
+    // NOTE validate reproInfo
+    if (!repoInfo) {
+      console.error(
+        `Found invalid GitHub URL: ${red(
+          `"${example}"`,
+        )}. Please fix the URL and try again.`,
+      );
+      process.exit(1);
+    }
+
+    // NOTE Do the repo exist?
+    if (!(await hasRepo(repoInfo))) {
+      console.error(
+        `Could not locate the repository for ${red(
+          `"${example}"`,
+        )}. Please check that the repository exists and try again.`,
+      );
+      process.exit(1);
+    }
+
+    return { example, defaultRef, repoInfo };
+  }
+
+  if (!(await existsInRepo(example, defaultRef))) {
+    console.error(
+      `Could not locate an example named ${red(
+        `"${example}"`,
+      )}. Please check that the example exists and try again.`,
+    );
+    process.exit(1);
+  }
+
+  return { example, defaultRef };
+}
+
+function isErrorLike(err: unknown): err is { message: string } {
+  return (
+    typeof err === 'object' &&
+    err !== null &&
+    typeof (err as { message?: unknown }).message === 'string'
+  );
+}
+
+export async function downloadAndExtract(
+  root: string,
+  { example, defaultRef, repoInfo }: ParsedExampleOption,
+): Promise<void> {
+  try {
+    if (repoInfo) {
+      console.log(
+        `Downloading files from repo ${cyan(example)}. This might take a moment.`,
+      );
+      console.log();
+      await downloadAndExtractRepo(root, repoInfo);
+    } else {
+      console.log(
+        `Downloading files for example ${cyan(example)}. This might take a moment.`,
+      );
+      console.log();
+      await downloadAndExtractExample(root, example, defaultRef);
+    }
+  } catch (err) {
+    // download error
+    throw new Error(isErrorLike(err) ? err.message : err + '');
```

Comment:
It doesn't seem doing anything good. If we want friendly error message we shouldn't throw anyway.

Reply:
nice your catch! I agree with you :)